components/server/src/authorization/relationship-updater.ts

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { ProjectDB, TeamDB, UserDB } from "@gitpod/gitpod-db/lib";
+import { AdditionalUserData, Organization, User } from "@gitpod/gitpod-protocol";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { inject, injectable } from "inversify";
+import { RedisMutex } from "../redis/mutex";
+import { Authorizer } from "./authorizer";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+
+@injectable()
+export class RelationshipUpdater {
+    public readonly version = 1;
+
+    constructor(
+        @inject(RedisMutex) private readonly mutex: RedisMutex,
+        @inject(UserDB) private readonly userDB: UserDB,
+        @inject(TeamDB) private readonly orgDB: TeamDB,
+        @inject(ProjectDB) private readonly projectDB: ProjectDB,
+        @inject(Authorizer) private readonly authorizer: Authorizer,
+    ) {}
+
+    /**
+     * Updates all relationships for a user according to the current state of the database.
+     * @param user
+     * @returns
+     */
+    public async migrate(user: User): Promise<User> {
+        if (user.additionalData?.fgaRelationshipsVersion === this.version) {
+            return user;
+        }
+        return this.mutex.using([`fga-migration-${user.id}`], 2000, async () => {
+            const before = new Date().getTime();
+
+            const updatedUser = await this.userDB.findUserById(user.id);
+            if (!updatedUser) {
+                throw new ApplicationError(ErrorCodes.NOT_FOUND, "User not found");
+            }
+            user = updatedUser;
+            if (user.additionalData?.fgaRelationshipsVersion === this.version) {
+                return user;
+            }
+            log.info({ userId: user.id }, `Updating FGA relationships for user.`, {
+                fromVersion: user?.additionalData?.fgaRelationshipsVersion,
+                toVersion: this.version,
+            });
+            const orgs = await this.orgDB.findTeamsByUser(user.id);
+
+            //TODO only remove relations that should no longer be there. Removing everything will break concurrent access.
+            await this.authorizer.removeAllRelationships("user", user.id);
+            for (const org of orgs) {
+                await this.authorizer.removeAllRelationships("organization", org.id);
+            }
+
+            // Add relationships
+            await this.updateUser(user);
+            for (const org of orgs) {
+                await this.updateOrganization(org);
+            }
+            AdditionalUserData.set(user, {
+                fgaRelationshipsVersion: this.version,
+            });
+            await this.userDB.updateUserPartial(user);
```

Comment:
Why not just update additionalData? :thinking:

Reply:
I don't understand this. Can you elaborate?

Reply:
To avoid "lost updates", especially when writing from potentially long(er)-running jobs, it's better to only update the minimal set of fields on the DB, instead of the whole `User` object:
```suggestion
            await this.userDB.updateUserPartial({ id: user.id, additionalUserData: user.additionalUserData });
```